scripts/quests/sandoria/Signed_in_Blood.lua

```diff
@@ -0,0 +1,190 @@
+                end,
+            },
+
+            onEventFinish = {
+                [1104] = function(player, csid, option, npc)
+                    quest:setVar(player, 'Prog', 2)
+                end,
+            },
+        },
+    },
+
+    -- Travel to Ordelle's Caves and hunt for an Ordelle Chest Key. Once you find a key, look for a Treasure Chest and open it.
+    -- Only people who have this quest flagged will receive some Torn-Out Pages (key item).
+    -- NOTE: this step is not included in this quest LUA. The key item is obtained via code in scripts/globals/treasure.lua
+
+    -- Return to Selbina and talk to the mayor. He'll reveal some information to you which will prove interesting to Sobane.
+    {
+        check = function(player, status, vars)
+            return status == QUEST_ACCEPTED and vars.Prog == 2
+        end,
+
+        [xi.zone.SOUTHERN_SAN_DORIA] = {
+            ['Sobane'] = {
+                onTrigger = function(player, npc)
+                    return quest:event(735)
+                end,
+            },
+        },
+
+        [xi.zone.SELBINA] = {
+            ['Abelard'] = {
+                onTrigger = function(player, npc)
+                    if player:hasKeyItem(xi.ki.TORN_OUT_PAGES) then
+                        return quest:progressEvent(1106)
+                    else
+                        return quest:event(1105)
+                    end
+                end,
+            },
+
+            onEventFinish = {
+                [1106] = function(player, csid, option, npc)
+                    quest:setVar(player, 'Prog', 3)
+                end,
+            },
+        },
+    },
+
+    -- Return to Sobane to complete this quest and to receive your reward.
+    {
+        check = function(player, status, vars)
+            return status == QUEST_ACCEPTED and vars.Prog == 3
+        end,
+
+        [xi.zone.SELBINA] = {
+            ['Abelard'] = {
+                onTrigger = function(player, npc)
+                    return quest:event(48)
+                end,
+            },
+        },
+
+        [xi.zone.SOUTHERN_SAN_DORIA] = {
+            ['Sobane'] = {
+                onTrigger = function(player, npc)
+                    return quest:progressEvent(736)
+                end,
+            },
+
+            onEventFinish = {
+                [736] = function(player, csid, option, npc)
+                    if quest:complete(player) then
+                        player:delKeyItem(xi.ki.TORN_OUT_PAGES)
+                        player:setLocalVar("signedInBloodNeedToZone", 1)
+                    end
+                end,
+            },
+        },
+    },
+
+    -- If you just completed Signed in Blood, you must zone out and back into Southern San d'Oria to flag Tea with a Tonberry
+    {
+        check = function(player, status, vars)
+            return status == QUEST_COMPLETED and player:getLocalVar("signedInBloodNeedToZone") == 1
+        end,
```

Comment:
Does this local var need to be handled after zoning so the player can progress?  Also, if the player logs out inbetween, I believe this section would be skipped.  Is that intended?

Reply:
That's a good question.  Because localVars clear on zone, using one like this can cover the "has the character zoned" question.  But with a localVar, it would also count logging out and back in as zoning, and I don't know if that's retail accurate.

So I've swapped out this logic in favor of player:needToZone().

That said, needToZone has its own gotcha, which is that it's not quest specific.  Consider this situation.

* Alice is doing two separate questlines in Selbina.
* In questline RED, there is a "must zone" requirement between quests RED-1 and RED-2 at Selbina NPC(R).
* In questline BLUE, there is a "must zone" requirement between quests BLUE-1 and BLUE-2 at Selbina NPC(B).
* Alice completes quest RED-1 and must now zone before acquiring RED-2 from the same NPC(R)
* Alice zones out, which clears her needToZone flag.
* Alice zones back in to Selbina.
* Before Alice acquires RED-2, she completes quest BLUE-1 at NPC(B), and must now zone before acquiring BLUE-2.
* She then returns to NPC(R), expecting to pick up RED-2, but now can't, because her needToZone flag has been set by the blue questline.

So using a localVar lets us handle the "needToZone" question distinctly within a given questline.  But it (probably) doesn't handle logout/login like retail.

While using needToZone (probably) handles logout/login like retail, but it uses a shared check across all questlines.

Reply:
Fair enough.  For complex situations like these, would it be someone that we should move to a charvar and clean up after?